### ISSUE TITLE
implement support for the NWB file format

### ISSUE THREAD

**Issue by robertoostenveld** [fieldtrip/fieldtrip#721]
See https://github.com/NeurodataWithoutBorders/matnwb for existing code. This would involve ft_filetype, ft_read_header, ft_read_data, ft_read_event, ft_read_spike. 

For testing I propose to use the public datasets at https://www.nwb.org/allen-cell-types-database/

@dorahermes and @choldgraf, could you perhaps look at this in relation to http://bit.ly/bids_ieeg?

**Comment by choldgraf** [comment]
This sounds great to me - would make me feel much better about officially supporting NWB in the iEEG spec. FieldTrip has far more users than `matnwb` so I think having export functionality in fieldtrip would go a long way.

**Comment by robertoostenveld** [comment]
I think the first goal would be to implement reading. If that works, writing could be considered.

**Comment by schoffelen** [comment]
Is this going to be followed up?

**Comment by robertoostenveld** [comment]
Since we don't have a direct benefit for this at the Donders, I don't see any of the local people implementing this. I also don't have time for this myself. Unless someone else picks this up, I propose that we close this issue.

**Comment by schoffelen** [comment]
OK, closing for now.

**Comment by rivkahrivkah** [comment]
Wow, my timing is great--I was just now looking to see if Fieldtrip supports NWB!

**Comment by schoffelen** [comment]
Uhm, Rebecca I just closed the issue. Do you mean that you want to reopen and self-assign it to push it forward?

**Comment by rivkahrivkah** [comment]
Sorry, my tone was off...I was just amused at the timing of my finding this issue, minutes after it was closed. I am new to both NWB and Fieldtrip so not really in the position to contribute, but maybe at a later date if we end going in that direction. Thanks for your efforts with Fieldtrip!

**Comment by schoffelen** [comment]
No worries, I was just probing whether I could enthuse you to take this forward... Or perhaps you know somebody in your close working environment who could take this on.

**Comment by dorahermes** [comment]
Just tagging Ben Dichter, who is in the NWB development team: @bendichter do you know anyone who may want to take this on?

**Comment by bendichter** [comment]
Thanks @dorahermes! I think there would be a lot of benefit for us to support FieldTrip. I've always been impressed with the level of documentation for this toolbox, and I think integration would provide value for our user base.

I'm looking through these functions and trying to think about how well the assumptions of this API maps onto the structure of NWB files, and I have some questions:

1. Generally speaking, a single NWB file contains *all* of the data for a single session. A single NWB file might contain the signal at the original sampling frequency as well as downsampled LFP data. If you just supply the filename, it won't be possible to specify which timeseries you mean. We could address this by allowing the user to specify the location of the timeseries or by trying to infer this.

1. In [ft_read_header](http://www.fieldtriptoolbox.org/reference/ft_read_header/):
We store data continuously, but we also store trial times, so we could return trialized data too. I don't see a flag for this though.

1. In [ft_read_spike](http://www.fieldtriptoolbox.org/reference/ft_read_spike/)
```matlab
spike.timestamp = 1xNchans cell-array, each element contains a vector (1 X Nspikes)
```
Does this imply that individual channels cannot share units? This assumption would not be appropriate for e.g. neuropixel or tetrode probes where multiple recordings sites often record from the same unit. Or I suppose we could just assign it to the channel where the waveform is maximum? Has this sort of thing been dealt with in another acquisition format already?

**Comment by schoffelen** [comment]
Thanks, @bendichter 

Perhaps some additional documentation can be found here: http://www.fieldtriptoolbox.org/tutorial/spike/
http://www.fieldtriptoolbox.org/tutorial/spikefield/

So to give some (partial) answers to your specific questions:
1. the 'mainstream' access to data in fieldtrip, is through ft_preprocessing, which relies on fileio/ft_read_data. This has been designed to deal with LFP data. Once people want to processe the spike data, they should do so through ft_read_spike. Thus, I would think that either calling ft_preprocessing (or ft_read_data) versus calling ft_read_spike on the same data file would lead to the lfp data or the spike information being processed.

2. the chunks of to-be-read data are specified as a so called 'trl' field in the configuration structure that is passed on to ft_preprocessing. ft_read_data subsequently does one of the following things: it either calls the low-level fileformat specific function only once to read in the full recording, and does the chunking afterwards, based on the information in cfg.trl, or it reads in the data chunk by chunk, if the low-level reading routine and the data format allows it. 
The information that represents the relevant events stored in a file are accessed through the ft_read_event function. Using a representation of these events, the ft_definetrial function can be used to extract this so-called trl matrix. There is ample documentation about this, e.g. http://www.fieldtriptoolbox.org/tutorial/preprocessing/

3. No, actually the design of the 'spike' structure in fieldtrip upon reading is to start off basic, i.e. have a vector of timestamps per contact, with associate peri-timestamp 'raw' waveforms. Reorganization into 'units' can be the next step. More details can be found in the documentation referred to above. Another lead can be the section on 'other types of data' here: http://www.fieldtriptoolbox.org/getting_started/

**Comment by bendichter** [comment]
@schoffelen thanks for the responses. I am definitely interested in this. I don't think I can currently support this on any of my current projects, but I will be keeping an eye open for opportunities.

**Comment by jens-k** [comment]
As part of next week's NWB hackathon, Tara van Viegen, Stefen Buergers and I are gonna attempt first steps on integrating NWB into Fieldtrip.

If anyone has any comments, tips, etc. please let us know. We have a preliminary plan on our project description page:
https://neurodatawithoutborders.github.io/nwb_hackathons/HCK08_2020_Remote/projects/fieldtrip_integration/

**Comment by jens-k** [comment]
We are looking for good fieldtrip example datasets as a gold standard to develop against. It would be great to have both continuous and trial datasets, both LFP and spike, and including a unit field.

The example from the two tutorials (p029_sort_final_01.nex) does not use the .unit field and seems to mix channels and units (judging from the label names).

**Comment by jens-k** [comment]
**Tara van Viegen (@TaravanViegen), Steffen Buergers (@sbuergers) and I wrote code to implement loading of NWB data (both LFP and spikes) with FieldTrip.
https://github.com/jens-k/fieldtrip/tree/nwb-integration**

The code requires the MatNWB toolbox with this extension: https://github.com/NeurodataWithoutBorders/matnwb/pull/215
(this should soon be part of NWB’s main branch).

It is assumed that the user is familiar with the basics of MatNWB ([Github MatNWB](https://neurodatawithoutborders.github.io/matnwb/)) and the user's system is in a state that would also allow loading the data using MatNWB (MatNWB must be in the path, generateCore must have been run, the correct schema must be installed, see [NWB schemas](https://github.com/NeurodataWithoutBorders/nwb-schema/releases)).

The code was tested with the following example datasets that are freely available online:
From https://gui.dandiarchive.org/#/file-browser/folder/5e6eb2b776569eb93f451f8d:
sub-YutaMouse20_ses-YutaMouse20-140324_behavior+ecephys.nwb
sub-YutaMouse41_ses-YutaMouse41-150831_behavior+ecephys.nwb 

From https://osf.io/hv7ja/ / https://gui.dandiarchive.org/#/dandiset/5e6ff4140fe0ec33e6b3d5d2:
P9HMH_NOID5.nwb

**Comment by jens-k** [comment]
At its current stage, the code is **not feature-complete**. For example:
- we have not implemented ft_read_event. NWB:N is a pretty generic dataformat and can contain very diverse types of data. So we were not sure how to programmatically and reliably create an event output that could be used in a trial function.
- waveforms are not read in by ft_read_spike
**Contributions are welcome!**

**Also further testing is welcome!** 
We only tested the code on a handful of datasets. NWB:N datasets can be very diverse and we would like to cover most common cases. Metadata can be saved at several places, data can be saved in objects of different classes (e.g., DataStub vs. DataPipe, with varying methods and properties). If a dataset cannot be read in by our code (while adhering to NWB:N 2.0 specs), we would like to know (please with link to the dataset, function used, and error stack).

Thanks!

Thanks!

**Comment by robertoostenveld** [comment]
Let me know how we can help move this forward. If the code is complete enough to be functional in some use cases, I propose to make a pull request and get it merged. There are many more data formats to which that applies. Better work with what we have and increment in small steps, rather than try to set the goals for a perfect but unattainable implementation.

The code changes in https://github.com/jens-k/fieldtrip/tree/nwb-integration seem clean. It would be good to have a `test_xxx.m` function, or an `inspect_xxx.m` function (when user interaction is needed, e.g. in properly installing NWB, generateCore, etc.)

**Comment by sbuergers** [comment]
I added a test_nwb.m in the test folder, which includes some instructions on how to get the data we used for testing and then run the provided code to convert it from NWB to fieldtrip. For us three this ran fine for the examples provided with the newest matNWB release (2.2.5). It also illustrates how the schema may need to be changes manually to the correct NWB schema version depending on the schema used for creating the file. 

I am making a pull request now.
